.github/workflows/release.yml

```diff
@@ -1,52 +1,41 @@
 name: Release
-
 on:
-  release:
-    types: [published]
-
+  push:
```

Comment:
I'm not overly happy about creating a new version every push. There are pushes where you don't need versioning (most of `docs`, `chore`, `style` kind of commit) and there are cases where it's better to include multiple breaking changes in only one major version. This library is fairly new, mainly worked by two people and we are already at version 2. Now that other people wants to contribute (big thank you btw) I'm concerned that by adding a new major version every breaking change it will spiral out of control.
Besides "convenience" is there any other reason for doing this?

Reply:
Yes there are alternatives @Pmyl.
From my experience, having a continuous deployment is very good.

**Solution 1 (bad design and not possible but I share it anyway):**
You could have a dedicated branch for example which is the stable version.
Nevertheless semantic-release try to create a standard on how we should handle versioning.
For example, the `branches` option allow only very specific branches.
So this branch would obviously be: master.

**Solution 2 (good by design - my favorite):**
Follow the Gitflow and add a develop branch which will become the entry point of your merge request (making it even the default branch).
That way, you can do what you want with code, make tons of breaking change and once you are ok to deliver it you merge it in master.

**Solution 3 (meh):**
Change the release workflow to no longer be automated.
Note that I recommend you to still create the release on a CI.
Then you should find a way to trigger the release by hand (never done that but I am pretty sure that you just can with a hack).
See this [feature request to allow manual trigger](https://github.community/t5/GitHub-Actions/GitHub-Actions-Manual-Trigger-Approvals/m-p/31517#M813).
See this [hack to do it](https://dev.to/s_abderemane/manual-trigger-with-github-actions-279e).

**Solution 4 (meh):**
Change the release workflow to no longer be automated.
Run the release by hand locally (should change the semantic option `ci` to false then).

Reply:
@Pmyl `docs`, `chore` and `style` types are configured to not trigger a release.
As I said, the configuration I provided allow you to define which types of commit you want to allow and which type of release it should trigger.

See the configuration:

```
"releaseRules":[
   {
      "type":"feat",
      "release":"minor"
   },
   {
      "type":"fix",
      "release":"patch"
   },
   {
      "type":"perf",
      "release":"patch"
   }
]
```

Reply:
I like solution 2!
But I'll investigate a bit more, I'll let you know asap, sorry

Reply:
I still think that there is nothing wrong about releasing every push, If we are pushing too many breaking changes and indeed creating too many versions that means that the library its not stable enough and the issue is somewhere else. As @C0ZEN pointed out we can configure the release rules to not trigger for chore, docs etc etc.

I'm also okay to move forward solution 2 with a manual process to merge into master when you are ready. However this is has a similar impact of what we currently have right now (Solution 3).

Reply:
trying to have an automated CD its not only useful to avoid manual process but its a mentality approach that will ensure that we produce quality code without hesitating. We should trust what we write and the infrastructure and tests that we run all the time. Numbers are free, if we reach version 20.0.1 its okay. Automated CD will improve our CI because we want to avoid mistakes.

Reply:
@Pmyl I understand your point and I'm relaxed with whatever solution we apply. Both of them have pro and cons in my view :)

Reply:
Automated CD works well on an application, a library is fundamentally different. CD on an application means seamless transition for the user, CD on a library means user having to update it potentially every hour instead of at fixed rate. I don't agree with `If we are pushing too many breaking changes [...] that means that the library its not stable enough`, there are many different reasons why breaking changes are needed, it doesn't always mean that the code is not stable. Change in requirements, minimum typescript version, also adding new supported type is in fact a breaking change, it may be that some users were relying on that being always undefined.

I like reading all the changes in a minor/major version of a package, it seems people is working on it. Having a version per bug/feature clutters the changelog and makes it less appealing to me.
e.g. If we approve 4 PRs in half an hour we get 4 new versions, I think it's worse than having one with the 4 commits written in it.

Going back to CD, even with CD you tend to work in sprints/timeboxes where at the end you have a "release note" with everything released. That is what I would like to have here as much as possible.

Let me think about it anyway

Reply:
Just so you know but I agree with @Pmyl.
A minimum of control of the releases is from my opinion required and you should be able to decide or not to release a new version.
That is why the Gitflow is pretty common nowadays and is a good fit for both your needs.
Only merging a branch to create a release: this is not seamless but this is not grunt work either.

Reply:
@Pmyl @C0ZEN Happy to go with solution 2, however I keep my mind on automated release on each merge to master for filtered commits messages. These are my reason :)

`CD on a library means user having to update it potentially every hour instead of at fixed rate.`
I see your point, it will create more noise? Yes. User don't have to update every time there is a change. they can just check the releases and decide to update or not.

` there are many different reasons why breaking changes are needed, it doesn't always mean that the code is not stable`

In this library we are following a policy of backwards compatibility for features, we haven't decided yet how long it should be. The point is that there shouldn't be too many.
I agree with you, for typescript updates we don't have a backwards compatibility functionality.


`I like reading all the changes in a minor/major version of a package, it seems people is working on it. Having a version per bug/feature clutters the changelog and makes it less appealing to me.`

I agree with you about readability. On the other hand you could say that it's easier to understand that one version change one thing instead of one version changed 50 things. 
About `it seems people is working on it`. If you release every day it looks like people are working on it anyway, event better, faster, more frequently. 


`if we approve 4 PRs in half an hour we get 4 new versions, I think it's worse than having one with the 4 commits written in it.`
In my opinion its better because as a consumer you can decide to only have one feature and not the entire block. A transition for a consumer its match smoother if they can update one bit at the time instead of plenty of things.